src/Microsoft.Health.Dicom.Core/Features/Common/ParallelEnumerable.cs

```diff
@@ -0,0 +1,117 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Dicom.Core.Features.Common;
+
+internal sealed class ParallelEnumerationOptions : ParallelOptions
+{
+    public int MaxBufferedItems { get; init; } = -1;
+}
+
+internal static class ParallelEnumerable
+{
+    [SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "Any failure should be surfaced to the reader.")]
+    public static async IAsyncEnumerable<TResult> SelectParallel<TSource, TResult>(
+        this IEnumerable<TSource> source,
+        Func<TSource, CancellationToken, ValueTask<TResult>> selector,
+        ParallelEnumerationOptions options)
+    {
+        // TODO: Support proper enumerator cancellation. Perhaps do not re-use ParallelOptions
+        var state = new State<TSource, TResult>(source, selector, options);
+        ValueTask producer = ProduceAsync(state);
+
+        await foreach (TResult item in state.Items.Reader.ReadAllAsync(options.CancellationToken))
+        {
+            yield return item;
+        }
+
+        await producer;
+
+        static async ValueTask ProduceAsync(State<TSource, TResult> state)
+        {
+            ChannelWriter<TResult> writer = state.Items.Writer;
+
+            try
+            {
+                // Note: The order is not deterministic.
+                // Items will be produced in the order that they are returned by the selector.
+                await Parallel.ForEachAsync(
```

Comment:
I'll emphasize in the comments that this procedure is unordered. But I think this is acceptable as the results are not paginated for WADO nor is ordered demanded by the spec: [10.4.1.1.2](https://dicom.nema.org/medical/dicom/current/output/html/part18.html#sect_10.4.1.1.2).

Of course, this may confusing to some consumers? I could refactor the code to enforce order if necessary.